Preprocess unicode escapes when parsing using JavaParser

kelloggm opened the pull request:
Fixes #6631.

Only the change at line 90 is actually required for the test to pass. However, it seemed like we'd want to use this option all the time, to match javac's behavior (JLS 3.3: javac removes all unicode escapes before it invokes the lexer). The downside of this change is that it could make error-reporting slightly less useful: problems in annotation files with unicode escapes will be reported at the "post-unicode preprocessing" location rather than the location before unicode preprocessing (which is a JavaParser limitation, and the reason that JavaParser has this option off by default even though it is the correct behavior from the JLS' perspective).

kelloggm commented:
The tests were failing because of the addition of the `preprocessUnicode` configuration option to the version of the `parse` method that takes a String as input. I suspect (but I am not sure) that the reason is that that version of the method is only called on strings that were produced originally by the other parse methods in the same class, so the problem was that it caused unicode escapes to be applied twice. That seems like a property of the way that these methods are currently used, though, and not the sort of detail that it's necessarily wise to rely on.

msridhar commented:
@kelloggm I can't quite understand your comment. Are you saying there are other methods in the class that do string transformations and preprocessing before parsing? I looked a bit but didn't see anything.

kelloggm commented:
> @kelloggm I can't quite understand your comment. Are you saying there are other methods in the class that do string transformations and preprocessing before parsing? I looked a bit but didn't see anything.

@msridhar sorry for being unclear. Let me explain more fully.

The relevant class is `JavaParserUtil` in `org.checkerframework.framework.util`. It provides a set of five static `parse*` methods that are intended as replacements for the methods in `StaticJavaParser`, because `StaticJavaParser` leaks memory:
1. `parseCompilationUnit(InputStream)`
2. `parseCompilationUnit(File)`
3. `parseCompilationUnit(String)`
4. `parseStubUnit(InputStream)`
5. `parseExpression(String expression, LanguageLevel languageLevel)`

(Note that there are a few other methods in the same class that provide wrappers for these, but these five implementations are the ones that actually contain code that configures and then invokes an instance of `JavaParser`.)

In my original version of this PR, I changed all five of these methods to add a call to `setPreprocessUnicodeEscapes(true)` on the relevant `ParserConfiguration` option. This caused the ajava tests run as part of the Nullness Checker's test suite to fail. (The logic for these tests is in `BaseTypeVisitor#testAnnotationInsertion()`; briefly, for each input file, they strip the annotations and then re-insert them and then check if the result is the same as the original input file.)

I identified that the reason that these tests failed when I added the unicode escape logic is that method 3 in my list above was producing a different result. In particular, `BaseTypeVisitor#testAnnotationInsertion()` makes two parse calls: one to one of the other methods in the list to parse the original file the first time, and then another to parse the string version of the code without annotations that's derived from the first (this is the call to 3 that I mentioned). This second call leads to different results if we also apply unicode escapes when we execute it, leading to the test failure.

Method 3 is only called internally by the Checker Framework (never to parse an external file), so I think it's safe to not do unicode escapes there. However, that's not really in the method's contract (but maybe it should be?), which was the reason for my original comment.

Please let me know if this needs more explanation.

mernst reviewed (approved):
Thanks!
